File: backend/app/schemas/doctor_note.py
from datetime import datetime

from typing import Any

from pydantic import BaseModel, ConfigDict, Field, model_validator

from backend.app.schemas.report import CLINICAL_DISCLAIMER


class DoctorNoteCreate(BaseModel):
    content: str | None = Field(default=None, min_length=1)
    title: str | None = None
    note_text: str | None = Field(default=None, min_length=1)
    note_type: str | None = None
    source_context: str | None = None
    active_ingredients: list[str] = Field(default_factory=list)
    drug_pair_keys: list[str] = Field(default_factory=list)
    diagnosis_keywords: list[str] = Field(default_factory=list)
    patient_tags: list[str] = Field(default_factory=list)
    applicability: dict[str, Any] = Field(default_factory=dict)
    priority: str | None = None

    @model_validator(mode="after")
    def require_text(self) -> "DoctorNoteCreate":
        content = (self.content or "").strip()
        note_text = (self.note_text or "").strip()
        if not content and not note_text:
            raise ValueError("content or note_text is required")
        if not self.content:
            self.content = note_text
        if not self.note_text:
            self.note_text = content
        return self


class DoctorNoteRead(BaseModel):
    id: int
    doctor_id: str
    content: str
    created_at: datetime
    disclaimer: str = CLINICAL_DISCLAIMER
    title: str | None = None
    note_text: str | None = None
    note_type: str | None = None
    source_context: str | None = None
    active_ingredients: list[str] = Field(default_factory=list)
    drug_pair_keys: list[str] = Field(default_factory=list)
    diagnosis_keywords: list[str] = Field(default_factory=list)
    patient_tags: list[str] = Field(default_factory=list)
    applicability: dict[str, Any] = Field(default_factory=dict)
    priority: str | None = None

    model_config = ConfigDict(from_attributes=True)


class DoctorNoteSearchItem(BaseModel):
    note_id: str | None = None
    title: str | None = None
    note_text: str | None = None
    note_type: str | None = None
    source_context: str | None = None
    active_ingredients: list[str] = Field(default_factory=list)
    drug_pair_keys: list[str] = Field(default_factory=list)
    diagnosis_keywords: list[str] = Field(default_factory=list)
    patient_tags: list[str] = Field(default_factory=list)
    applicability: dict[str, Any] = Field(default_factory=dict)
    priority: str | None = None
    score: float = 0.0
    match_reason: str | None = None
    created_at: str | None = None

File: frontend/src/types/doctorNote.ts
export type DoctorMemoryNote = {
  note_id?: string
  id?: string | number
  title?: string | null
  note_text?: string | null
  content?: string | null
  note_type?: string | null
  source_context?: string | null
  active_ingredients?: string[] | null
  drug_pair_keys?: string[] | null
  diagnosis_keywords?: string[] | null
  patient_tags?: string[] | null
  applicability?: Record<string, unknown> | null
  priority?: string | null
  score?: number | null
  match_reason?: string | null
  created_at?: string | null
  is_newly_saved?: boolean
}

export type DoctorMemoryResponse = {
  matched_notes: DoctorMemoryNote[]
  warnings?: string[] | null
}

export type DoctorNoteCreatePayload = {
  content: string
  note_text: string
  title?: string
  note_type?: string
  source_context: 'prescription_audit'
  active_ingredients: string[]
  drug_pair_keys: string[]
  diagnosis_keywords: string[]
  patient_tags: string[]
  applicability: Record<string, unknown>
  priority: 'normal' | 'high'
}
